core/trino-main/src/main/java/io/trino/sql/planner/DomainTranslator.java

```diff
@@ -525,35 +523,18 @@ protected ExtractionResult visitComparisonExpression(ComparisonExpression node,
          */
         private Optional<NormalizedSimpleComparison> toNormalizedSimpleComparison(Map<NodeRef<Expression>, Type> expressionTypes, ComparisonExpression comparison)
         {
-            Object left = new IrExpressionInterpreter(comparison.getLeft(), plannerContext, session, expressionTypes).optimize(NoOpSymbolResolver.INSTANCE);
-            Object right = new IrExpressionInterpreter(comparison.getRight(), plannerContext, session, expressionTypes).optimize(NoOpSymbolResolver.INSTANCE);
-
+            Expression left = comparison.getLeft();
+            Expression right = comparison.getRight();
             Type leftType = expressionTypes.get(NodeRef.of(comparison.getLeft()));
             Type rightType = expressionTypes.get(NodeRef.of(comparison.getRight()));
 
-            checkArgument(leftType.equals(rightType), "left and right type do not match in comparison expression (%s)", comparison);
-
-            if (left instanceof Expression == right instanceof Expression) {
-                // we expect one side to be expression and other to be value.
-                return Optional.empty();
-            }
-
-            Expression symbolExpression;
-            ComparisonExpression.Operator comparisonOperator;
-            NullableValue value;
-
-            if (left instanceof Expression) {
-                symbolExpression = comparison.getLeft();
-                comparisonOperator = comparison.getOperator();
-                value = new NullableValue(rightType, right);
-            }
-            else {
-                symbolExpression = comparison.getRight();
-                comparisonOperator = comparison.getOperator().flip();
-                value = new NullableValue(leftType, left);
-            }
-
-            return Optional.of(new NormalizedSimpleComparison(symbolExpression, comparisonOperator, value));
+            return switch (left) {
+                case Constant constant when !(right instanceof Constant) ->
+                        Optional.of(new NormalizedSimpleComparison(right, comparison.getOperator().flip(), new NullableValue(leftType, constant.getValue())));
+                case Expression expression when right instanceof Constant constant ->
+                        Optional.of(new NormalizedSimpleComparison(left, comparison.getOperator(), new NullableValue(rightType, constant.getValue())));
+                default -> Optional.empty();
```

Comment:
It's complicated because it's trying to detect the case when one is a constant and the other is not, but the fact that Constant <: Expression makes it hard to represent. I think I'm going to take a different approach entirely. It's not very readable.

Reply:
Updated.